src/DevBetterWeb.Infrastructure/Services/AlumniGraduationService.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DevBetterWeb.Core;
+using DevBetterWeb.Core.Entities;
+using DevBetterWeb.Core.Interfaces;
+using DevBetterWeb.Infrastructure.Identity.Data;
+using Microsoft.AspNetCore.Identity;
+
+namespace DevBetterWeb.Infrastructure.Services
+{
+  public class AlumniGraduationService : IAlumniGraduationService
+  {
+    private const int DAYS_IN_TWO_YEARS = 365 * 2;
+
+    private readonly IUserLookupService _userLookupService;
+    private readonly IRepository<Member> _repository;
+    private readonly IGraduationCommunicationsService _graduationCommunicationsService;
+    private readonly UserManager<ApplicationUser> _userManager;
+
+    public AlumniGraduationService(IUserLookupService userLookupService,
+      IRepository<Member> repository,
+      IGraduationCommunicationsService graduationCommunicationsService,
+      UserManager<ApplicationUser> userManager)
+    {
+      _userLookupService = userLookupService;
+      _repository = repository;
+      _graduationCommunicationsService = graduationCommunicationsService;
+      _userManager = userManager;
+    }
+
+    public async Task GraduateMembersIfNeeded(AppendOnlyStringList messages)
+    {
+      var members = await _repository.ListAsync();
```

Comment:
Do you need all members? Would just a list of non-alumni members be more efficient?

Reply:
Yes, that would be more efficient. But I don't think I can really define that in a spec because I have to go from member to its userId to the corresponding user to that user's roles. Which would be a lot of lambda expressions that I don't think a spec can handle. If I'm wrong on that I could change it to that though